### DISCUSSION TITLE
Hello

### DISCUSSION THREAD

**Discussion by popstarfreas** [spicylobstergames/astratomic#86] [General]
Hi! I was just looking around at similar projects in the same space as something I'm working on and found yours.

I've built my project in Rust using Bevy ECS and Vulkano. Example video:
https://github.com/user-attachments/assets/eba60fe4-7c8c-4d9e-a080-1973462ebf1c

The actual falling sand part isn't visible in this video because the video was originally intended to show off some user interface and animation changes. But it seems we have similar inspirations. I've taken inspiration from Noita, Terraria and Wizard of Legend. My game is currently closed source, but it's interesting to see what others are working on that have similarities. Especially when there's discussions to be had about how things are implemented or how to overcome certain technical issues. I've particularly been torn on how to use the GPU more to lower the CPU load from the simulation and lighting engine that are currently CPU bound. I've noticed not much activity recently on this repo, are you still working on this?

**Comment by Zac8668** [comment]
You project is really cool!
As of right now I'm not working on astratomic 😔 unfortunately uni takes a lot of time, but I plan to work on it again in the future.

About the simulation stuff I think it's possible to do it on gpu but probably very limited, I think I remember someone that was able to make it work.
The lighting for sure can be made 100% GPU but I didn't got to it yet.
The only stuff of the simulation I tought about doing on the GPU is fluid simulation, I had some ideas and I'm pretty sure it's possible.

Edit: But to make the simulation more fast you for sure need to have some multithreading and chunk system going on, so that you can update multiple parts of the world at the same time, if you haven't implemented this yet, it's totally worth it.

**Comment by Zac8668** [comment]
Also, if you didn't yet, check out PieKing's project, he does the lighting on the GPU side and some other incredible things.
https://github.com/PieKing1215/FallingSandEngine

**Comment by popstarfreas** [comment]
Thanks!
I know the feeling hah. I started my project in Summer 2022 and have had some gaps in dev because of lack of time.

Yeah, I managed to make the simulation multithreaded with some careful use of unsafe rust, and got a chunk system and a grid similar to what the noita dev talks showed, but I'm still noticing some performance issues most likely when loading a new chunk that has a bunch of water flowing from where it got placed in world gen. I'm currently trying to target 120 tps so that the game can render up to 120 fps without the sim starting to tick at a lower rate than frame rate, but that gives me only 8.3 ms per update not including any rendering getting in the way.
I did try implementing sound for things like water like their talks showed but so far it's hit and miss because sometimes the sound is averaging out at some place that doesn't have any water for example. I'll have to fire up noita again and see if I can get any more info on how the sound position works.
I also have some artifacts from how I've tried to speed up the flow of some things like water and sand that I need to get rid of:
![image](https://forum.dark-gaming.com/uploads/default/original/3X/b/f/bff06b65246de0fcaaab212de14c8c09e2548bc9.webp)
Because by default they seemed to flow quite slowly they way I implemented them, but that could be because I've not seen specific approaches beyond the basic movement demonstrated in the noita dev talks.

My lighting is currently based on Terraria's approach, which is at least multithreaded and allows light to travel around corners.
This is an example of how a dark area looks when illuminated by a torch:
![image](https://github.com/user-attachments/assets/5584bff2-b1e9-4984-8694-b5bcb76bc1f8)

In my searching I've not yet seen someone actually describe Terraria's approach specifically, because it does lend itself to at least being multithreaded, but it is CPU based. I'm just trying to see if I can make something that looks similar to that style of lighting (because my game also uses blocks) but try to do it on the GPU, since that would give more time for the CPU to do the simulation. I have been looking into some shaders in different projects but not 100% sure on the details yet since I've not tried to write one yet, that's what I was looking at today. I did notice that Noita has some shaders for their fire, which I've not entirely understood yet how it works because it would be nice to implement something similar, have you looked into this yet?

For my project I think the sim itself has to be CPU-based because I plan to add networking and I've heard that GPU results are not deterministic, but doing it on the GPU does sound cool! And I guess bevy ecs which I'm using isn't technically deterministic either, so that's another thing I have to figure out.

Oh yeah, that repo is what I looked at before this one, and the link on its README is how I got here. I'll be interested to see how your project develops when you get time again. It seems like there's people that got inspired by Noita but I've not yet seen anything beyond projects in the early stages. And it's cool to see how other people approach some of the challenges of building something that includes this type of simulation.
